src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,343 @@
+/*
+MIT License
+
+Copyright(c) 2020 Futurewei Cloud
+
+    Permission is hereby granted,
+    free of charge, to any person obtaining a copy of this software and associated documentation files(the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and / or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions :
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#include "LogStream.h"
+#include <k2/common/Chrono.h>
+#include <k2/config/Config.h>
+#include <k2/dto/Collection.h>
+#include <k2/dto/Persistence.h>
+#include <k2/transport/RPCDispatcher.h>
+#include <k2/transport/RPCTypes.h>
+#include <k2/transport/Status.h>
+#include <k2/transport/TXEndpoint.h>
+#include <k2/dto/ControlPlaneOracle.h>
+#include <k2/dto/MessageVerbs.h>
+#include <k2/dto/LogStream.h>
+#include <cstdlib>
+#include <cctype>
+#include <algorithm>
+#include <random>
+
+namespace k2 {
+
+LogStream::LogStream() {
+    K2INFO("dtor");
+    _switched = false;    
+}
+
+LogStream::~LogStream() {
+    K2INFO("~dtor");
+}
+
+seastar::future<> 
+LogStream::init(String cpo_url, String persistenceClusrerName){
+    K2INFO("LogStream Init");
+    
+    _cpo = CPOClient(cpo_url);
+    return _client.init(persistenceClusrerName, cpo_url);
+}
+
+seastar::future<>
+LogStream::create(String logStreamName){
+    if (_logStreamCreate){
+        K2INFO("Log streaam already created");
+        throw k2::dto::LogStreamExistError("Log stream already created");
+    }
+
+    _logStreamName = std::move(logStreamName);
+    _logStreamCreate = true;

```

Comment:
I think you can only set _logStreamCreate to true at the last step inside this function, otherwise, there maybe client try to write into this LogStream even before you create any meta plogs, or wal plogs below.

Reply:
Set the _logStreamCreate=true at the penultimate step. The last step of the creation is to write the WAL Plog id to metadata Plog, which requires the _logStreamCreate==true